modules/images/image-loading-optimization/class-ilo-grouped-url-metrics.php

```diff
@@ -0,0 +1,295 @@
+	 * @return array LCP elements keyed by its minimum viewport width. If there is no supported LCP element at a breakpoint, then `false` is used.
+	 */
+	public function get_lcp_elements_by_minimum_viewport_widths(): array {
+		$lcp_element_by_viewport_minimum_width = array();
+		foreach ( $this->groups as $viewport_minimum_width => $breakpoint_url_metrics ) {
+
+			// The following arrays all share array indices.
+			$seen_breadcrumbs   = array();
+			$breadcrumb_counts  = array();
+			$breadcrumb_element = array();
+
+			foreach ( $breakpoint_url_metrics as $breakpoint_url_metric ) {
+				foreach ( $breakpoint_url_metric->get_elements() as $element ) {
+					if ( ! $element['isLCP'] ) {
+						continue;
+					}
+
+					$i = array_search( $element['xpath'], $seen_breadcrumbs, true );
+					if ( false === $i ) {
+						$i                       = count( $seen_breadcrumbs );
+						$seen_breadcrumbs[ $i ]  = $element['xpath'];
+						$breadcrumb_counts[ $i ] = 0;
+					}
+
+					$breadcrumb_counts[ $i ] += 1;
+					$breadcrumb_element[ $i ] = $element;
+					break; // We found the LCP element for the URL metric, go to the next URL metric.
+				}
+			}
+
+			// Now sort by the breadcrumb counts in descending order, so the remaining first key is the most common breadcrumb.
+			if ( $seen_breadcrumbs ) {
+				arsort( $breadcrumb_counts );
+				$most_common_breadcrumb_index = key( $breadcrumb_counts );
+
+				$lcp_element_by_viewport_minimum_width[ $viewport_minimum_width ] = $breadcrumb_element[ $most_common_breadcrumb_index ];
+			} elseif ( ! empty( $breakpoint_url_metrics ) ) {
+				$lcp_element_by_viewport_minimum_width[ $viewport_minimum_width ] = false; // No LCP image at this breakpoint.
+			}
+		}
+
+		// Now merge the breakpoints when there is an LCP element common between them.
+		$prev_lcp_element = null;
+		return array_filter(
+			$lcp_element_by_viewport_minimum_width,
+			static function ( $lcp_element ) use ( &$prev_lcp_element ) {
+				$include = (
+					// First element in list.
+					null === $prev_lcp_element
+					||
+					( is_array( $prev_lcp_element ) && is_array( $lcp_element )
+						?
+						// This breakpoint and previous breakpoint had LCP element, and they were not the same element.
+						$prev_lcp_element['xpath'] !== $lcp_element['xpath']
+						:
+						// This LCP element and the last LCP element were not the same. In this case, either variable may be
+						// false or an array, but both cannot be an array. If both are false, we don't want to include since
+						// it is the same. If one is an array and the other is false, then do want to include because this
+						// indicates a difference at this breakpoint.
+						$prev_lcp_element !== $lcp_element
+					)
+				);
+				$prev_lcp_element = $lcp_element;
+				return $include;
+			}
+		);
+	}
+
+	/**
+	 * Checks whether all groups have URL metrics.
+	 *
+	 * @return bool Whether all groups have URL metrics.
+	 */
+	public function are_all_groups_populated(): bool {
```

Comment:
Maybe the name could also be simplified. Maybe we could call it `is_populated()`?

Separately, I would recommend we add another method here called `is_complete()`: This would return true if all groups are not just populated but have enough metrics to satisfy the sample size. This could be used in at least one place below in favor of the more complicated current implementation of `get_needed_minimum_viewport_widths()`.

Reply:
I've renamed it to `is_every_group_populated()` in 3523e98b. I think this is more clear than something shorter like `is_populated()` because that could mean is _any_ group populated.